### ISSUE TITLE
eval CODEREF should work

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8708]
Migrated from [rt.perl.org#41066](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=41066) (status was 'rejected')

Searchable as RT41066$

**Comment by p5pRT** [comment]
### From @ilmari

#### Created by @ilmari



eval EXPR always stringifies EXPR\, causing the following error passed a
coderef&#8203;:

$ perl \-wle 'eval sub \{\}; warn $@&#8203; if $@&#8203;;'
Undefined subroutine &main&#8203;::CODE called at \(eval 1\) line 1\.

I'd suggest checking of EXPR is a coderef and treating it like eval
BLOCK\, i\.e\. doing the equivalent of

&nbsp;    eval \{ &$ref; \}

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.8:

Configured by Debian Project at Thu Dec  7 13:58:37 UTC 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.15.7, archname=i486-linux-gnu-thread-multi
    uname='linux vernadsky 2.6.15.7 #1 smp sat sep 30 10:21:42 utc 2006 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.8 -Dsitearch=/usr/local/lib/perl/5.8.8 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.8 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.1.2 20061115 (prerelease) (Ubuntu 4.1.1-20ubuntu1)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.5.so, so=so, useshrplib=true, libperl=libperl.so.5.8.8
    gnulibc_version='2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /etc/perl
    /usr/local/lib/perl/5.8.8
    /usr/local/share/perl/5.8.8
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    /usr/local/lib/perl/5.8.7
    /usr/local/share/perl/5.8.7
    .


Environment for perl v5.8.8:
    HOME=/home/ilmari
    LANG=en_GB.UTF-8
    LANGUAGE=en_GB:en
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ilmari/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From guest@guest.guest.xxxxxxxx

On Sun Dec 10 11&#8203;:05&#8203;:26 2006\, ilmari wrote&#8203;:
> eval EXPR always stringifies EXPR\, causing the following error passed a
> coderef&#8203;:
> 
> $ perl \-wle 'eval sub \{\}; warn $@&#8203; if $@&#8203;;'
> Undefined subroutine &main&#8203;::CODE called at \(eval 1\) line 1\.
> 
> I'd suggest checking of EXPR is a coderef and treating it like eval
> BLOCK\, i\.e\. doing the equivalent of
> 
>     eval \{ &$ref; \}

This is bad\, for the same reason that a regular sub checking
ref\($somearg\) to decide what to do is bad\.

Right now\, this&#8203;:
&nbsp;   perl \-we'use overload q\!""\! => sub \{"1"\}; $coderef = bless sub\{"0"\};
print eval $coderef'
is unambiguous; let's not change that\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From a.r.ferreira@gmail.com

On 12/10/06\, via RT Dagfinn Ilmari Mannsåker \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Dagfinn Ilmari Mannsåker
> \# Please include the string&#8203;:  \[perl \#41066\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=41066 >
>
>
>
> This is a bug report for perl from ilmari@&#8203;ilmari\.org\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
>
> eval EXPR always stringifies EXPR\, causing the following error passed a
> coderef&#8203;:
>
> $ perl \-wle 'eval sub \{\}; warn $@&#8203; if $@&#8203;;'
> Undefined subroutine &main&#8203;::CODE called at \(eval 1\) line 1\.
>
> I'd suggest checking of EXPR is a coderef and treating it like eval
> BLOCK\, i\.e\. doing the equivalent of
>
>         eval \{ &$ref; \}

\[The following is my personal assessment of the presented issue\.\]

I am not sure it makes sense\. The 'eval' already has two meanings&#8203;:

&nbsp;    eval STRING;   \# compile and run the code in the string \(trapping
exceptions\)
&nbsp;    eval BLOCK;    \# run the code \(trapping exceptions\)

Your proposal would add a third one&#8203;:

&nbsp;    eval CODEREF; \# run the code ref \(trapping exceptions\)

In the example\, you gave\, you want

&nbsp;        eval sub \{\}

to work\. But

&nbsp;        eval \{\}

already works and is shorter\. You probably had in mind invoking
coderefs in variables such as the piece you mentioned after&#8203;:

&nbsp;       eval \{ &$ref; \}

If you write the above or

&nbsp;       eval \{ $ref\->\(\) \}

\(which has the extra benefit of being easily extended with a call with
more arguments\)\, you will be fine\. The shorter form

&nbsp;       eval $ref;

would add \(as I said above\) only a third meaning which looks like \->\(\)
but which traps the exceptions\. I think it is too little to be gained
by adding extra complexity for something that can be done easily in
other ways\.

Yet another thing to remember is that distinguishing between 'eval
STRING' and 'eval CODEREF' would be done at runtime\, while the current
distinction on the two forms may happen at compile time\. So it is
subtler and depends on what ends as the result of the expression EXPR
of the documented form 'eval EXPR'\.

For the record\, Perl 6 made these two eval meanings into different
constructions\, now using \(IIRC\)&#8203;:

&nbsp;     eval STRING;
&nbsp;     try   BLOCK;

Now let us wait for the porters to tell their words on the subject\.

Thanks for contributing to the discussion and improvement of the Perl
interpreter\.

Adriano Ferreira\.

> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=low
> \-\-\-
> Site configuration information for perl v5\.8\.8&#8203;:
>
> Configured by Debian Project at Thu Dec  7 13&#8203;:58&#8203;:37 UTC 2006\.
>
> Summary of my perl5 \(revision 5 version 8 subversion 8\) configuration&#8203;:
>   Platform&#8203;:
>     osname=linux\, osvers=2\.6\.15\.7\, archname=i486\-linux\-gnu\-thread\-multi
>     uname='linux vernadsky 2\.6\.15\.7 \#1 smp sat sep 30 10&#8203;:21&#8203;:42 utc 2006 i686 gnulinux '
>     config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-Dcccdlflags=\-fPIC \-Darchname=i486\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.8 \-Darchlib=/usr/lib/perl/5\.8 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.8\.8 \-Dsitearch=/usr/local/lib/perl/5\.8\.8 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Uusesfio \-Uusenm \-Duseshrplib \-Dlibperl=libperl\.so\.5\.8\.8 \-Dd\_dosuid \-des'
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
>     useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>     use64bitint=undef use64bitall=undef uselongdouble=undef
>     usemymalloc=n\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
>     optimize='\-O2'\,
>     cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
>     ccversion=''\, gccversion='4\.1\.2 20061115 \(prerelease\) \(Ubuntu 4\.1\.1\-20ubuntu1\)'\, gccosandvers=''
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>     alignbytes=4\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='cc'\, ldflags =' \-L/usr/local/lib'
>     libpth=/usr/local/lib /lib /usr/lib
>     libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
>     perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
>     libc=/lib/libc\-2\.5\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.8\.8
>     gnulibc\_version='2\.5'
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
>     cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib'
>
> Locally applied patches&#8203;:
>
>
> \-\-\-
> @&#8203;INC for perl v5\.8\.8&#8203;:
>     /etc/perl
>     /usr/local/lib/perl/5\.8\.8
>     /usr/local/share/perl/5\.8\.8
>     /usr/lib/perl5
>     /usr/share/perl5
>     /usr/lib/perl/5\.8
>     /usr/share/perl/5\.8
>     /usr/local/lib/site\_perl
>     /usr/local/lib/perl/5\.8\.7
>     /usr/local/share/perl/5\.8\.7
>     \.
>
> \-\-\-
> Environment for perl v5\.8\.8&#8203;:
>     HOME=/home/ilmari
>     LANG=en\_GB\.UTF\-8
>     LANGUAGE=en\_GB&#8203;:en
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=/home/ilmari/bin&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/sbin&#8203;:/bin&#8203;:/usr/bin/X11&#8203;:/usr/games
>     PERL\_BADLANG \(unset\)
>     SHELL=/bin/bash
>
>

**Comment by p5pRT** [comment]
### From @rgarcia

On 12/12/06\, Adriano Ferreira \<a\.r\.ferreira@&#8203;gmail\.com> wrote&#8203;:
> I am not sure it makes sense\. The 'eval' already has two meanings&#8203;:
>
>     eval STRING;   \# compile and run the code in the string \(trapping
> exceptions\)
>     eval BLOCK;    \# run the code \(trapping exceptions\)
>
> Your proposal would add a third one&#8203;:
>
>     eval CODEREF; \# run the code ref \(trapping exceptions\)

Actuallly\, those two meanings are parsed differently \(different
grammar rules in perly\.y\) and compiled differently \(different opcodes&#8203;:
entereval and entertry\)\.

> Yet another thing to remember is that distinguishing between 'eval
> STRING' and 'eval CODEREF' would be done at runtime\, while the current
> distinction on the two forms may happen at compile time\. So it is
> subtler and depends on what ends as the result of the expression EXPR
> of the documented form 'eval EXPR'\.

Exactly\. And since we have different compiled forms\, we would have to
add special runtime cases\, to switch from one op to another\.

Basically\, I think \(like Adriano\) that it's quite a lot of complexity
put for just a small amount of syntactic sugar\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'rejected'
